File: cardbattlemodule/src/main/java/com/syahputrareno975/cardbattlemodule/task/RoomStreamTask.java
package com.syahputrareno975.cardbattlemodule.task;

import android.os.AsyncTask;
import cardBattle.CardBattle;
import cardBattle.cardBattleServiceGrpc;
import com.syahputrareno975.cardbattlemodule.interfaces.RoomStreamEvent;
import com.syahputrareno975.cardbattlemodule.interfaces.RoomStreamEventController;
import com.syahputrareno975.cardbattlemodule.interfaces.SimpleUnit;
import com.syahputrareno975.cardbattlemodule.model.NetworkConfig;
import com.syahputrareno975.cardbattlemodule.model.card.CardModel;
import com.syahputrareno975.cardbattlemodule.model.player.PlayerModel;
import com.syahputrareno975.cardbattlemodule.model.playerWithCard.PlayerWithCardsModel;
import com.syahputrareno975.cardbattlemodule.model.room.RoomDataModel;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.stub.StreamObserver;
import java.util.concurrent.TimeUnit;

import static cardBattle.CardBattle.roomStream.EventCase.EXCMESSAGE;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toAllPlayerBattleResultModelModel;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toCardModelGRPC;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toEndResultModel;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toPlayerModelGRPC;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toPlayerWithCardsModel;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toPlayerWithCardsModelGRPC;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toRoomModel;
import static com.syahputrareno975.cardbattlemodule.model.ModelCasting.toRoomModelGRPC;

public class RoomStreamTask extends AsyncTask<Void,Void,Boolean>{

    private ManagedChannel channel;
    private RoomDataModel roomData ;
    private PlayerWithCardsModel player ;
    private NetworkConfig net;
    private RoomStreamEvent roomStreamEvent;
    private Throwable failed = null;

    private RoomStreamEventController controller;
    private StreamObserver<CardBattle.roomStream> request;
    private StreamObserver<CardBattle.roomStream> response;
    private String error  = "";


    public RoomStreamTask( PlayerWithCardsModel player,RoomDataModel roomData, NetworkConfig net, RoomStreamEvent roomStreamEvent) {
        this.roomData = roomData;
        this.player = player;
        this.net = net;
        this.roomStreamEvent = roomStreamEvent;
    }


    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        this.channel = ManagedChannelBuilder.forAddress(this.net.Url, this.net.Port)
                .usePlaintext(true)
                .build();

        response = new StreamObserver<CardBattle.roomStream>() {
            @Override
            public void onNext(CardBattle.roomStream value) {
                holder.event = value;
                publishProgress();
            }

            @Override
            public void onError(Throwable t) {
                failed = t;
                error += t.getMessage();
                holder.stop = true;
            }

            @Override
            public void onCompleted() {
                holder.stop = true;
            }

        };

        controller = new RoomStreamEventController() {
            @Override
            public void playerJoin(PlayerModel p) {
                // left this empty
                // because its already send when connected
            }

            @Override
            public void getOneRoom(String id) {
                request.onNext(CardBattle.roomStream.newBuilder()
                        .setIdRoom(roomData.Id)
                        .setGetOneRoom(toRoomModelGRPC(roomData))
                        .build());
            }

            @Override
            public void deployCard(PlayerModel p, CardModel c) {
                request.onNext(CardBattle.roomStream.newBuilder()
                        .setIdRoom(roomData.Id)
                        .setDeployCard(CardBattle.playerAndCard.newBuilder()
                                .setCardData(toCardModelGRPC(c))
                                .setClient(toPlayerModelGRPC(p))
                                .build())
                        .build());
            }

            @Override
            public void pickUpCard(PlayerModel p, CardModel c) {
                request.onNext(CardBattle.roomStream.newBuilder()
                        .setIdRoom(roomData.Id)
                        .setPickupCard(CardBattle.playerAndCard.newBuilder()
                                .setCardData(toCardModelGRPC(c))
                                .setClient(toPlayerModelGRPC(p))
                                .build())
                        .build());
            }

            @Override
            public void leftGame(PlayerModel p, RoomDataModel r, SimpleUnit action) {
                holder.actionHolder = action;
                request.onNext(CardBattle.roomStream.newBuilder()
                        .setIdRoom(roomData.Id)
                        .setPlayerLeft(toPlayerWithCardsModelGRPC(player))
                        .build());
            }
        };
    }

    Holder holder = new Holder();
    @Override
    protected Boolean doInBackground(Void... voids) {
        try {
            cardBattleServiceGrpc.cardBattleServiceStub stub = cardBattleServiceGrpc.newStub(this.channel);
            request = stub.cardBattleRoomStream(response);

            request.onNext(CardBattle.roomStream.newBuilder()
                    .setIdRoom(roomData.Id)
                    .setPlayerJoin(toPlayerWithCardsModelGRPC(player))
                    .build());


            roomStreamEvent.onConnected(controller);

            while (!holder.stop){
                // infinite loop
            }

            request.onCompleted();


        } catch (Exception e){
            error += e.getMessage();
        }

        return true;
    }

    @Override
    protected void onProgressUpdate(Void... values) {
        super.onProgressUpdate(values);
        if (holder.event != null){
            switch (holder.event.getEventCase()){
                case COUNTDOWN :
                    roomStreamEvent.onCountDown(holder.event.getCountDown().getBattleTime(), toRoomModel(holder.event.getCountDown().getUpdatedRoom()));
                    break;

                case PLAYERJOIN:
                    roomStreamEvent.onPlayerJoin(toPlayerWithCardsModel(holder.event.getPlayerJoin()).Owner);
                    break;

                case PLAYERLEFT:
                   PlayerModel p = toPlayerWithCardsModel(holder.event.getPlayerLeft()).Owner;
                    holder.stop = p.Id.equals(player.Owner.Id);
                    holder.left = p.Id.equals(player.Owner.Id);
                    roomStreamEvent.onPlayerLeft(p);
                    break;

                case ONROOMUPDATE:
                    roomStreamEvent.onRoomUpdate(toRoomModel(holder.event.getOnRoomUpdate()));
                    break;

                case BATTLERESULT :
                    roomStreamEvent.onBattleResult(toAllPlayerBattleResultModelModel(holder.event.getBattleResult()));
                    break;

                case RESULT:
                    holder.stop = true;
                    roomStreamEvent.onResult(toEndResultModel(holder.event.getResult()));
                    break;

                case ONDRAW:
                    int flag = holder.event.getOnDraw();
                    holder.stop = flag == 1 || flag == 2 || flag == 3;
                    roomStreamEvent.onDraw(flag);
                    break;

                case DEPLOYCARD:

                    // left this empty
                    break;

                case PICKUPCARD:

                    // left this empty
                    break;

                case GETONEROOM:
                    roomStreamEvent.onGetRoomData(toRoomModel(holder.event.getGetOneRoom()));
                    break;

                case EXCMESSAGE:
                    roomStreamEvent.onException(holder.event .getExcMessage().getExceptionMessage(), holder.event .getExcMessage().getExceptionFlag());
                    break;

                default:
                    break;
            }
        }
    }

    @Override
    protected void onPostExecute(Boolean aBoolean) {
        super.onPostExecute(aBoolean);
        try {
            this.channel.shutdown().awaitTermination(5, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } catch (NullPointerException e) {
            throw new RuntimeException(e);
        }

        if (!error.equals("")) {
            roomStreamEvent.onError(error);
            return;
        }

        if (holder.left) {
            holder.actionHolder.invoke();
        }

        roomStreamEvent.onDisconnected();
    }

    class Holder {
        public SimpleUnit actionHolder = new SimpleUnit() {
            @Override
            public void invoke() {

            }
        };
        public Boolean left = false;
        public Boolean stop = false;
        public CardBattle.roomStream event = null;
    }
}

File: app/src/main/java/com/syahputrareno975/simplecardbattle/task/RoomStreamTask.kt
package com.syahputrareno975.simplecardbattle.task

import android.os.AsyncTask
import cardBattle.CardBattle
import cardBattle.cardBattleServiceGrpc
import com.syahputrareno975.simplecardbattle.interfaces.LobbyStreamController
import com.syahputrareno975.simplecardbattle.interfaces.RoomStreamEvent
import com.syahputrareno975.simplecardbattle.interfaces.RoomStreamEventController
import com.syahputrareno975.simplecardbattle.model.ModelCasting
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toAllPlayerBattleResultModelModel
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toAllPlayerBattleResultModelModelGRPC
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toAllPlayerWithCardsModelGRPC
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toCardModelGRPC
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toEndResultModel
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toPlayerModel
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toPlayerModelGRPC
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toPlayerWithCardsModel
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toPlayerWithCardsModelGRPC
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toRoomModel
import com.syahputrareno975.simplecardbattle.model.ModelCasting.Companion.toRoomModelGRPC
import com.syahputrareno975.simplecardbattle.model.NetworkConfig
import com.syahputrareno975.simplecardbattle.model.card.CardModel
import com.syahputrareno975.simplecardbattle.model.player.PlayerModel
import com.syahputrareno975.simplecardbattle.model.playerWithCard.PlayerWithCardsModel
import com.syahputrareno975.simplecardbattle.model.room.RoomDataModel
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.stub.StreamObserver
import java.util.concurrent.TimeUnit

class RoomStreamTask : AsyncTask<Void,Void,Boolean>{

    lateinit var channel: ManagedChannel
    var roomData : RoomDataModel
    var player : PlayerWithCardsModel
    var net: NetworkConfig
    var roomStreamEvent :  RoomStreamEvent
    private var failed: Throwable? = null

    lateinit var controller : RoomStreamEventController
    lateinit var request : StreamObserver<CardBattle.roomStream>
    lateinit var response : StreamObserver<CardBattle.roomStream>
    var error: String = ""

    constructor(player : PlayerWithCardsModel,roomData : RoomDataModel,net: NetworkConfig,roomStreamEvent : RoomStreamEvent) : super() {
        this.player = player
        this.roomData = roomData
        this.roomStreamEvent = roomStreamEvent
        this.net = net
    }

    override fun onPreExecute() {
        super.onPreExecute()
        this.channel = ManagedChannelBuilder.forAddress(this.net.Url, this.net.Port)
            .usePlaintext(true)
            .build()


        response = object : StreamObserver<CardBattle.roomStream> {
            override fun onNext(value: CardBattle.roomStream?) {
                holder.event = value
                publishProgress()
            }

            override fun onError(t: Throwable?) {
                failed = t
                error += t!!.message
                holder.stop = true
            }

            override fun onCompleted() {
                holder.stop = true
            }
        }

        controller = object : RoomStreamEventController {
            override fun playerJoin(p: PlayerModel) {
                // left this empty
                // because its already send when connected
            }

            override fun getOneRoom(id: String) {
                request.onNext(CardBattle.roomStream.newBuilder()
                    .setIdRoom(roomData.Id)
                    .setGetOneRoom(toRoomModelGRPC(roomData))
                    .build())
            }

            override fun deployCard(p: PlayerModel, c: CardModel) {
                request.onNext(CardBattle.roomStream.newBuilder()
                    .setIdRoom(roomData.Id)
                    .setDeployCard(CardBattle.playerAndCard.newBuilder()
                        .setCardData(toCardModelGRPC(c))
                        .setClient(toPlayerModelGRPC(p))
                        .build())
                    .build())
            }

            override fun pickUpCard(p: PlayerModel, c: CardModel) {
                request.onNext(CardBattle.roomStream.newBuilder()
                    .setIdRoom(roomData.Id)
                    .setPickupCard(CardBattle.playerAndCard.newBuilder()
                        .setCardData(toCardModelGRPC(c))
                        .setClient(toPlayerModelGRPC(p))
                        .build())
                    .build())
            }

            override fun leftGame(p: PlayerModel, r: RoomDataModel) {
                request.onNext(CardBattle.roomStream.newBuilder()
                    .setIdRoom(roomData.Id)
                    .setPlayerLeft(toPlayerWithCardsModelGRPC(player))
                    .build())
            }
        }
    }

    val holder = Holder()

    override fun doInBackground(vararg params: Void?): Boolean {
        try {
            val stub = cardBattleServiceGrpc.newStub(this.channel)
            request = stub.cardBattleRoomStream(response)

            request.onNext(CardBattle.roomStream.newBuilder()
                .setIdRoom(roomData.Id)
                .setPlayerJoin(toPlayerWithCardsModelGRPC(player))
                .build())


            roomStreamEvent.onConnected(controller)

            while (!holder.stop){
                // infinite loop
            }

            request.onCompleted()


        } catch (e : Exception){
            error += e.message
        }

        return true
    }

    override fun onProgressUpdate(vararg values: Void?) {
        super.onProgressUpdate(*values)
        if (holder.event != null){
            when (holder.event!!.eventCase){
                cardBattle.CardBattle.roomStream.EventCase.COUNTDOWN -> {
                    roomStreamEvent.onCountDown(holder.event!!.countDown)
                }
                cardBattle.CardBattle.roomStream.EventCase.PLAYERJOIN -> {
                    roomStreamEvent.onPlayerJoin(toPlayerWithCardsModel(holder.event!!.playerJoin).Owner)
                }
                cardBattle.CardBattle.roomStream.EventCase.PLAYERLEFT -> {
                    val p = toPlayerWithCardsModel(holder.event!!.playerLeft).Owner
                    holder.stop = p.Id == player.Owner.Id
                    holder.left = p.Id == player.Owner.Id
                    roomStreamEvent.onPlayerLeft(p)
                }
                cardBattle.CardBattle.roomStream.EventCase.ONROOMUPDATE -> {
                    roomStreamEvent.onRoomUpdate(toRoomModel(holder.event!!.onRoomUpdate))
                }
                cardBattle.CardBattle.roomStream.EventCase.BATTLERESULT -> {
                    roomStreamEvent.onBattleResult(toAllPlayerBattleResultModelModel(holder.event!!.battleResult))
                }
                cardBattle.CardBattle.roomStream.EventCase.RESULT -> {
                    holder.stop = true
                    roomStreamEvent.onResult(toEndResultModel(holder.event!!.result))
                }
                cardBattle.CardBattle.roomStream.EventCase.ONDRAW -> {
                    holder.stop = holder.event!!.onDraw == 1 || holder.event!!.onDraw == 2
                    roomStreamEvent.onDraw(holder.event!!.onDraw)
                }
                cardBattle.CardBattle.roomStream.EventCase.DEPLOYCARD -> {

                    // left this empty

                }
                cardBattle.CardBattle.roomStream.EventCase.PICKUPCARD -> {

                    // left this empty

                }
                cardBattle.CardBattle.roomStream.EventCase.GETONEROOM -> {
                    roomStreamEvent.onGetRoomData(toRoomModel(holder.event!!.getOneRoom))
                }
                cardBattle.CardBattle.roomStream.EventCase.EXCMESSAGE -> {
                    roomStreamEvent.onException(holder.event!!.excMessage.exceptionMessage,holder.event!!.excMessage.exceptionFlag)
                }
                else -> {

                }
            }
        }
    }

    override fun onPostExecute(result: Boolean?) {
        super.onPostExecute(result)
        try {
            this.channel.shutdown().awaitTermination(5, TimeUnit.SECONDS)
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        } catch (e: NullPointerException) {
            throw RuntimeException(e)
        }

        if (error != "") {
            roomStreamEvent.onError(error)
            return
        }

        if (holder.left) {
            roomStreamEvent.onLeft()
            return
        }

        roomStreamEvent.onDisconnected()

    }

    class Holder {
        var left = false
        var stop = false
        var event : CardBattle.roomStream? = null
    }
}